### PR TITLE
Code refactoring

### PR DESCRIPTION
**Pull request by varun-andhra-mapup**
- Refactored geometric predicates in joins.
- Removed conversion of NBI coordinates using formula.
- Added fuzzy matching for road and stream names.
- Resolved bridge points at the same location due to same stream intersections.
- Added filtering for tunnel=culverts using osmium.

### CODE REVIEW

**Comment by maneesh-mapup** [comment]
Add ChatGPT / Claude research as a comment

**Comment by varun-andhra-mapup** [comment]
Hey guys, can you please provide your feedback on this PR? Let me know if you need anything.

@parshu-mapup @tanishq-sharma-mapup

**Comment by parshu-mapup** [comment]
## Code Review Feedback

### Strengths
1. **Modular Design:** Functions are well-defined and serve specific purposes, aiding readability and maintenance.
2. **Documentation:** Function docstrings provide a clear explanation of each function's purpose, inputs, and outputs.
3. **Use of QGIS Processing Algorithms:** Effectively leverages QGIS's processing capabilities for complex geospatial tasks.
4. **Error Handling:** Basic checks are in place to verify the validity of loaded layers, preventing the script from proceeding with invalid data.

### Areas for Improvement
1. **Error Handling and Logging:**
   - Consider implementing more robust error handling with specific exceptions and error messages.
   - Use logging instead of print statements for better monitoring and debugging, especially for long-running processes.

2. **Code Efficiency:**
   - Review the use of in-memory layers (`"memory:"`). If the data volume is large, this could lead to performance issues.
   - Where possible, streamline the processing workflow to minimize repeated operations on large datasets.

3. **Data Management:**
   - Ensure that all temporary files and layers are properly managed and deleted after use to prevent clutter and resource consumption.

4. **Parameterization:**
   - Consider making file paths, buffer distances, and other constants configurable via a configuration file or command-line arguments. This increases the flexibility of the script.

5. **Code Comments and Structure:**
   - Add more inline comments to explain complex logic, especially in sections involving spatial analysis and joins.
   - Break down larger functions into smaller, more manageable pieces if they become too complex.

6. **Performance Considerations:**
   - For processing large datasets, explore parallel processing options or batch processing to improve performance.
   - Evaluate whether all fields and features being processed are necessary, potentially reducing data size early in the workflow.

7. **Testing:**
   - Implement unit tests for critical functions, particularly those involving data manipulation and processing, to ensure accuracy and robustness.

8. **Output Management:**
   - Include more informative output messages and possibly summary statistics about the processed data (e.g., number of features processed, number of layers created).
   - Ensure output files are consistently named and stored in a structured manner to facilitate later retrieval and use.

Overall, the script demonstrates strong foundational work in processing and analyzing geospatial data with QGIS. Addressing these areas can further improve the robustness, performance, and maintainability of the code.

**Comment by parshu-mapup** [comment]
## Summary of Code Functionality

This Python script processes geospatial data using the QGIS framework. The primary objectives include filtering and analyzing bridge and culvert data from OpenStreetMap (OSM) and National Bridge Inventory (NBI) sources. The key functionalities are:

1. **Layer Loading and Initialization:**
   - Initializes QGIS, loads vector layers from specified file paths, and prepares them for processing.

2. **Data Processing Functions:**
   - **`create_buffer`:** Generates buffer zones around vector features.
   - **`filter_osm_data`:** Applies filter expressions to isolate specific features in OSM data.
   - **`explode_osm_data`:** Processes OSM data to expand fields stored in a key-value format.

3. **Spatial Analysis and Joins:**
   - **`join_by_location`:** Joins attributes between layers based on spatial relationships.
   - **`filter_nbi_layer`:** Filters NBI layer features, excluding those identified in previous steps.

4. **CSV and GeoPackage Outputs:**
   - Exports processed layers to CSV and GeoPackage formats for further analysis and use.

5. **Specialized Processing:**
   - **`process_bridge`, `process_layer_tag`, `process_parallel_bridges`, `process_nearby_bridges`, `process_culverts_from_pbf`:** Specific workflows for handling different types of bridge and culvert data, including filtering, buffering, and joining.

6. **Comprehensive Processing Pipeline:**
   - The `main` function orchestrates the entire workflow, sequentially processing layers and data to produce the final output.

This script is used for analyzing geospatial data related to infrastructure, specifically bridges and culverts, by leveraging QGIS's geospatial processing capabilities.

**Comment by tanishq-sharma-mapup** [comment]
1. [processing-scripts/04-obtaining-bridge-coordinates/01-obtain-bridge-split-info.py](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/11/commits/91188153679e4098d9c7188feec5fb081b81becd#diff-df96a7d1b0a692fc6e5912d3707d5c88f0d79c8be0e3d432c37e60928fea3c39) script uses print statement for logging. Please use a logging mechanism to create a log file with different levels of logs (info, warn, error).

2. Many of the scripts that consume files and produce outputs use constant filenames throughout. Please create constants at the start or end of the script for specifying all input and output file names.

3. [processing-scripts/03-associating-data/03-get-point-projections-on-ways.py](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/11/files?diff=split&w=0#diff-5499315b4d391733fe9b23bc455c7fd5d2c0cf4ed087b288912e69f590370268) lacks error handling and logging, especially for string-to-int conversions, such as when calculating "final_osm_id". Please add error handling to catch conversion errors.

4. [processing-scripts/04-obtaining-bridge-coordinates/01-obtain-bridge-split-info.py](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/11/commits/91188153679e4098d9c7188feec5fb081b81becd#diff-df96a7d1b0a692fc6e5912d3707d5c88f0d79c8be0e3d432c37e60928fea3c39) line 147. The for loop continues even after finding and processing the required way_id. Please consider breaking the loop when the required way is found and processed.

ChatGPT Link of research: https://chatgpt.com/share/9482c050-d507-4f2c-9999-a06d767136e0

File: processing-scripts/01-filtering-data/02-process-filter-nbi-bridges.py [hunk 1]
```diff
@@ -84,49 +37,34 @@ def convert_to_gpkg(df, output_gpkg_file):
...
+# input_csv = "input-data/Kentucky-NBI-bridge-data.csv"
+input_csv = "input-data/NBI-Kentucky-bridge-data.csv"
 output_duplicate_exclude_csv = (
     "output-data/csv-files/Kentucky-bridge-chosen-coordinates.csv"
 )
 output_gpkg_file = "output-data/gpkg-files/NBI-Kentucky-Bridge-Data.gpkg"
-process_coordinates(
-    input_csv, output_convert_csv, output_duplicate_exclude_csv, output_gpkg_file
-)
+process_coordinates(input_csv, output_duplicate_exclude_csv, output_gpkg_file)
```
Comment (maneesh-mapup) [inline processing-scripts/01-filtering-data/02-process-filter-nbi-bridges.py:70]: @varun-andhra-mapup do we have AI code review added to this? If yes, where are the AI suggestions? Also, you need to make a change to start the code review process

Comment (varun-andhra-mapup) [reply processing-scripts/01-filtering-data/02-process-filter-nbi-bridges.py:70]: @maneesh-mapup We have the PR AI Agent installed in this repo. However, this PR is 2 weeks old and has already been closed and merged into main. I am creating a new PR now with some logging changes, where the PR Agent will run by itself.

**Pull request merged**